miden-crypto/src/hash/algebraic_sponge/rescue/rpo/mod.rs

```diff
@@ -168,32 +183,265 @@ impl Rpo256 {
+    pub const RATE_RANGE: Range<usize> = Rpo256::RATE_RANGE;
+
+    /// The capacity portion of the state is located in elements 0, 1, 2, and 3.
+    pub const CAPACITY_RANGE: Range<usize> = Rpo256::CAPACITY_RANGE;
+
+    /// The output of the hash function can be read from state elements 4, 5, 6, and 7.
+    pub const DIGEST_RANGE: Range<usize> = Rpo256::DIGEST_RANGE;
+
+    // RESCUE PERMUTATION
+    // --------------------------------------------------------------------------------------------
+
+    /// Applies RPO permutation to the provided state.
+    ///
+    /// This delegates to the RPO256 implementation which applies 7 rounds of the
+    /// Rescue Prime Optimized permutation.
+    #[inline(always)]
+    pub fn apply_permutation(state: &mut [Felt; STATE_WIDTH]) {
+        Rpo256::apply_permutation(state);
     }
+}
 
-    fn merge(values: &[Self::Digest; 2]) -> Self::Digest {
-        <Self as AlgebraicSponge>::merge(values)
+// PLONKY3 TRAIT IMPLEMENTATIONS
+// ================================================================================================
+
+impl Permutation<[Felt; STATE_WIDTH]> for RpoPermutation256 {
+    fn permute_mut(&self, state: &mut [Felt; STATE_WIDTH]) {
+        Self::apply_permutation(state);
     }
+}
 
-    fn merge_many(values: &[Self::Digest]) -> Self::Digest {
-        <Self as AlgebraicSponge>::merge_many(values)
+impl CryptographicPermutation<[Felt; STATE_WIDTH]> for RpoPermutation256 {}
+
+// TYPE ALIASES FOR PLONKY3 INTEGRATION
+// ================================================================================================
+
+/// RPO-based hasher using Plonky3's PaddingFreeSponge.
+///
+/// This provides a sponge-based hash function with:
+/// - WIDTH: 12 field elements (total state size)
+/// - RATE: 8 field elements (input/output rate)
+/// - OUT: 4 field elements (digest size)
+pub type RpoHasher = PaddingFreeSponge<RpoPermutation256, 12, 8, 4>;
+
+/// RPO-based compression function using Plonky3's TruncatedPermutation.
+///
+/// This provides a 2-to-1 compression function for Merkle tree construction with:
+/// - CHUNK: 2 (number of input chunks - i.e., 2 digests of 4 elements each = 8 elements)
+/// - N: 4 (output size in field elements)
+/// - WIDTH: 12 (total state size)
+///
+/// The compression function takes 8 field elements (2 digests) as input and produces
+/// 4 field elements (1 digest) as output.
+pub type RpoCompression = TruncatedPermutation<RpoPermutation256, 2, 4, 12>;
+
+/// RPO-based challenger using Plonky3's DuplexChallenger.
+///
+/// This provides a Fiat-Shamir transform implementation for interactive proof protocols,
+/// with:
+/// - F: Generic field type (typically the same as Felt)
+/// - WIDTH: 12 field elements (sponge state size)
+/// - RATE: 8 field elements (rate of absorption/squeezing)
+pub type RpoChallenger<F> = DuplexChallenger<F, RpoPermutation256, 12, 8>;
+
+#[cfg(test)]
+mod p3_tests {
```

Comment:
nit: shouldn't these tests go into the `src/hash/algebraic_sponge/rescue/rpo/tests.rs` module?